# Najbolj popularni animeji
## Projektna naloga pri predmetu Programiranje 1.

Za to projektno nalogo sem zajel prvih 3000 najbolj popularnih animejev s strani [MyAnimeList](https://myanimelist.net/topanime.php). Podatki so rangirani glede na [tale](https://myanimelist.net/info.php?go=topanime) kriterij.
Za pridobitev in obdelavo podatkov sem, bolj kot ne, uporabil kodo, ki jo je profesor Pretnar pripravil za predavanja.

#### Naloga je sestavljena iz štirih delov:
1. ZBRANI PODATKI:
    - Definicija animeja
    - Delovne hipoteze
    - Z zajemom sestavljena tabela.
2. OSNOVNA STATISTIKA: Podatki, pridobljeni s preureditvijo tabele.
3. IZRAČUNANA STATISTIKA: Podatki, ki jih moramo izračunati oz. razbrati, odvisnosti in podobno.
4. DODATNE HIPOTEZE: za zabavo bom morda pogledal še kakšno hipotezo, ki je nisem napovedal v Readme.md datoteki.



### ZBRANI PODATKI
* Podatki ki sem jih zbral za posamezni anime:
    - rang
    - id
    - naslov
    - leto izida
    - dolžina(število epizod)
    - tip
    - MyAnimeList ocena
    - število ogledov

Glavni razlog, zakaj nisem zajel več različnih tipov podatkov, je dejstvo, da če bi želel zajeti več podatkov (npr. studio/ji, ki je/so sproduciral/i anime), bi moral za vsaki anime posebej dol potegniti spletno stran (torej 3000 spletnih strani). V prvi vrsti dvomim, da bi mi strežnik to dopustil v enem poskusu, kaj šele v vseh (mnogih) poskusih, ki sem jih izvedel, da sem sestavil to nalogo, v drugi vrsti pa je to res ogromna količina datotek, ki bi jih moral prečesati in v primeru napake, prečesati še enkrat. Krajše povedano: Časovna zahtevnost projekta bi se znatno povečala. 

#### Kaj je anime?
[Anime](https://en.wikipedia.org/wiki/Anime) je originalno japonski izraz za animacijo, torej vse oblike animiranih medijev (filmi, serije itd.). Izven Japonske se termin nanaša specifično na animacije, ki izvirajo iz Japonske, ali pa na animacijski, ki se je najprej razširil na Japonskem, za katerega je značilna tako vizualna kot tudi karakterna barvitost/pestrost in fantazijske teme. Spletna stran kot anime upošteva vse vzhodno-azijske animacije, ki so bodisi japonske, bodisi spadajo pod prej omenjeni stil.
Animeje se deli na več tipov, za nas so pomembni tile:
* TV: Animirana serija, ki je bila (najprej) predvajana po televiziji.
* Movie: Animiran film, ki je bil (najprej) predvajan v kinih.
* Special: Posebej izdana animacija, ki se jo lahko najde na televiziji, ali pa je izdana kot DVD.
* OVA: "O(riginal) V(ideo) A(nimation)", animacija, ki je direktno izdana na disku in predhodno ni predvajana na televiziji ali v kinih. 
* ONA: "O(riginal) N(etwork) A(nimation)", animacija, ki je najprej izdana na spletu.
O slednjih treh tipih si lahko več preberete [tukaj](https://aminoapps.com/c/anime/page/blog/defining-ova-ona-special-episodes-and-movies/bEto_u8PomgzmRaW7MLjQxz5LoZ6VP)

Kot sem prej omenil, je rang animeja določen s pomočjo kriterija s strani MyAnimeList. Tako bom na začetku malce pokomentiral ta kriterij.
Za izračun "Težnostnega ranga" animeja potrebujemo povprečno oceno tega animeja, število vseh ocen animeja, minimalno število ocen, da se sploh "splača" računat (trenutno je to 50 ocen) in povprečno oceno po vseh animejih. Anime lahko ocenijo samo registrirani uporabniki te spletne strani, kar pomeni, da podatki, ki so bili zbrani s te strani, niso nujno najbolj natančni, še posebej ne za starejše animeje.

Skozi vso projektno nalogo bom odgovarjal na sledeča vprašanja:
* Ali je zlata doba animejev že mimo, ali se šele začenja?
* Ali so bolj popularni animirani filmi, ali serije?
* Kako dolžina (TV)serije vpliva na njen uspeh?
* V kakšnem razmerju sta število ogledov in MyAnimeList ocena?

In [39]:
%matplotlib inline

import pandas as pd
import matplotlib
from matplotlib import pyplot as plt

pd.options.display.max_rows = 20

animeji = pd.read_csv('obdelani-podatki/vsi-animeji.csv', index_col = 'rang')

In [40]:
animeji

id                                             naslov      tip  \
rang                                                                      
1      5114                   Fullmetal Alchemist: Brotherhood       TV   
2     32281                                     Kimi no Na wa.    Movie   
3      9253                                        Steins;Gate       TV   
4     28977                                           Gintama°       TV   
5     11061                             Hunter x Hunter (2011)       TV   
6       820                               Ginga Eiyuu Densetsu      OVA   
7      9969                                           Gintama'       TV   
8     35180                         3-gatsu no Lion 2nd Season       TV   
9     15417                                Gintama': Enchousen       TV   
10    28851                                     Koe no Katachi    Movie   
...     ...                                                ...      ...   
2991    478                                 Sousei no Aquarion       TV   
2992   2452                              Space Adventure Cobra    Movie   
2993  35774                         Starmyu 2nd Season Special  Special   
2994   3332                                 Stop!! Hibari-kun!       TV   
2995  21881                                Sword Art Online II       TV   
2996   4532                          Taiyou no Yuusha Fighbird       TV   
2997  20035  Toaru Majutsu no Index Movie: Endymion no Kise...    Movie   
2998   7654                          Tokyo Magnitude 8.0 Recap  Special   
2999   2577                               Tom Sawyer no Bouken       TV   
3000  31370                               Tonkatsu DJ Agetarou       TV   

      st_epizod  leto   ogledi  ocena  
rang                                   
1            64  2009  1327051   9.24  
2             1  2016   866945   9.15  
3            24  2011  1116166   9.14  
4            51  2015   226008   9.13  
5           148  2011   818148   9.11  
6           110  1988   140128   9.10  
7            51  2011   235640   9.10  
8            22  2017   114343   9.05  
9            13  2012   137790   9.05  
10            1  2016   621825   9.03  
...         ...   ...      ...    ...  
2991         26  2005    57573   7.17  
2992          1  1982     5258   7.17  
2993          1  2017     1649   7.17  
2994         35  1983     2996   7.17  
2995         24  2014   911220   7.17  
2996         48  1991     1746   7.17  
2997          1  2013    15645   7.17  
2998          1  2009     5668   7.17  
2999         49  1980     3648   7.17  
3000         12  2016     9884   7.17  

[2930 rows x 7 columns]

Iz do sedaj neznanega razloga ni bilo zajetih 70 datotek. Zaradi tega av tabeli ni 3000 vrstic, temveč zgolj 2930. Napako bom poskusil odpraviti, če mi to ne uspe, pa je zajeti vzorec še vedno dovolj obsežen, da ta izguba ni občutna.

### OSNOVNA STATISTIKA
Oglejmo si, kateri izmed zajetih animejev je najstarejši(posledično bomo videli tudi kateri je najmlajši) in kateri je bil do zdaj najbolj gledan.

In [41]:
animeji_po_starosti = animeji.sort_values('leto')

In [42]:
animeji_po_starosti

id                                             naslov      tip  \
rang                                                                      
2235  22419                                       Acchi Kocchi    Movie   
2700   2747                                      Tetsuwan Atom       TV   
2310   3009                                         Tiger Mask       TV   
2710   5760                               Dororo to Hyakkimaru       TV   
227    2402                                      Ashita no Joe       TV   
1041   1412                                          Lupin III       TV   
1961   2253                                         Mazinger Z       TV   
1861    501                                           Doraemon       TV   
2561   5658                                     Great Mazinger       TV   
2550   2225                               Alps no Shoujo Heidi       TV   
...     ...                                                ...      ...   
1282  36094                                 Hakumei to Mikochi       TV   
1285  37475                           Kishuku Gakkou no Juliet       TV   
287   37171                                      Asobi Asobase       TV   
2410  37764                              Mobile Suit Gundam NT    Movie   
2408  36616  Mob Psycho 100: Reigen - Shirarezaru Kiseki no...  Special   
1321  37341                                Yuru Camp△ Specials  Special   
1323  34443                                               Baki      ONA   
2399  37210     Isekai Maou to Shoukan Shoujo no Dorei Majutsu       TV   
1244  36513                         Rokuhoudou Yotsuiro Biyori       TV   
2565  34420      Isekai Izakaya: Koto Aitheria no Izakaya Nobu      ONA   

      st_epizod  leto  ogledi  ocena  
rang                                  
2235          1  1962    5528   7.35  
2700        193  1963    7128   7.24  
2310        105  1969    5871   7.34  
2710         26  1969    2455   7.23  
227          79  1970   34916   8.31  
1041         23  1971   51718   7.73  
1961         92  1972    9012   7.42  
1861         26  1973   11674   7.44  
2561         56  1974    3011   7.27  
2550         52  1974   17853   7.27  
...         ...   ...     ...    ...  
1282         12  2018   32738   7.63  
1285         12  2018   81223   7.63  
287          12  2018  102621   8.24  
2410          1  2018    4041   7.31  
2408          1  2018   29388   7.31  
1321          3  2018   19356   7.62  
1323         26  2018   35551   7.61  
2399         12  2018  180050   7.31  
1244         12  2018   15341   7.65  
2565         24  2018   12861   7.27  

[2930 rows x 7 columns]

Iz urejene tabele vidimo, da je najstarejši zajeti anime animiran film "Acchi Kocchi" iz leta 1962, najmlajši anime pa ONA "Isekai Izakaya: Koto Aitheria no Izakaya Nobu", ki je izšel leta 2018. Iz povezave v definiciji animeja pa vemo, da je prvi izšel že leta 1917, kar potrjuje, da je naš vzorec nepopoln.

In [43]:
animeji_po_ogledih = animeji.sort_values('ogledi', ascending=False)

In [44]:
animeji_po_ogledih

id                              naslov      tip  st_epizod  leto  \
rang                                                                        
53     1535                          Death Note       TV         37  2006   
113   16498                  Shingeki no Kyojin       TV         25  2013   
1410  11757                    Sword Art Online       TV         25  2012   
1      5114    Fullmetal Alchemist: Brotherhood       TV         64  2009   
45    30276                       One Punch Man       TV         12  2015   
581   22319                         Tokyo Ghoul       TV         12  2014   
3      9253                         Steins;Gate       TV         24  2011   
253    6547                        Angel Beats!       TV         13  2010   
177   19815                     No Game No Life       TV         12  2014   
716      20                              Naruto       TV        220  2002   
...     ...                                 ...      ...        ...   ...   
1941   9340                         Cheburashka    Movie          1  2010   
2963   9200  Dae Jang Geum: Jang Geum's Dream 2       TV         26  2007   
2340  11835    Qin Shi Ming Yue: Zhu Zi Bai Jia       TV         34  2010   
1527  36910               Wu Geng Ji 2nd Season      ONA         42  2017   
2375  15905  Qin Shi Ming Yue: Wanli Changcheng       TV         37  2012   
2990   2758               Shippuu! Iron Leaguer       TV         52  1993   
2165  36561            Doupo Cangqiong Specials  Special          2  2017   
2984  29936    Qin Shi Ming Yue: Junlin Tianxia      ONA         75  2014   
2548  38250                 Xiao Lu He Xiao Lan      ONA         72  2018   
2590  38491                      Xing Chen Bian      ONA         12  2018   

       ogledi  ocena  
rang                  
53    1582700   8.66  
113   1472352   8.48  
1410  1415367   7.59  
1     1327051   9.24  
45    1160132   8.71  
581   1123638   7.99  
3     1116166   9.14  
253   1089044   8.29  
177   1074522   8.39  
716   1073428   7.90  
...       ...    ...  
1941      781   7.42  
2963      769   7.17  
2340      720   7.33  
1527      715   7.55  
2375      707   7.32  
2990      707   7.17  
2165      697   7.37  
2984      501   7.17  
2548      437   7.28  
2590      374   7.27  

[2930 rows x 7 columns]

Iz te tabele pa lahko razberemo, da je do zdaj najbolj ogledan anime bil "Death Note" iz leta 2006, najmanj ogledov pa je prejel "Xing Chen Bian" iz 2018.

### IZRAČUNANA STATISTIKA
V tem delu naloge bom s pomočjo grafov odgovoril na zgoraj zastavljena vprašanja.
Najprej grupiramo animeje po desetletjih, da bodo grafi bolj pregledni, tam kjer bo to imelo kak efekt.

In [45]:
animeji['desetletje'] = 10 * (animeji.leto // 10)

In [46]:
animeji

id                                             naslov      tip  \
rang                                                                      
1      5114                   Fullmetal Alchemist: Brotherhood       TV   
2     32281                                     Kimi no Na wa.    Movie   
3      9253                                        Steins;Gate       TV   
4     28977                                           Gintama°       TV   
5     11061                             Hunter x Hunter (2011)       TV   
6       820                               Ginga Eiyuu Densetsu      OVA   
7      9969                                           Gintama'       TV   
8     35180                         3-gatsu no Lion 2nd Season       TV   
9     15417                                Gintama': Enchousen       TV   
10    28851                                     Koe no Katachi    Movie   
...     ...                                                ...      ...   
2991    478                                 Sousei no Aquarion       TV   
2992   2452                              Space Adventure Cobra    Movie   
2993  35774                         Starmyu 2nd Season Special  Special   
2994   3332                                 Stop!! Hibari-kun!       TV   
2995  21881                                Sword Art Online II       TV   
2996   4532                          Taiyou no Yuusha Fighbird       TV   
2997  20035  Toaru Majutsu no Index Movie: Endymion no Kise...    Movie   
2998   7654                          Tokyo Magnitude 8.0 Recap  Special   
2999   2577                               Tom Sawyer no Bouken       TV   
3000  31370                               Tonkatsu DJ Agetarou       TV   

      st_epizod  leto   ogledi  ocena  desetletje  
rang                                               
1            64  2009  1327051   9.24        2000  
2             1  2016   866945   9.15        2010  
3            24  2011  1116166   9.14        2010  
4            51  2015   226008   9.13        2010  
5           148  2011   818148   9.11        2010  
6           110  1988   140128   9.10        1980  
7            51  2011   235640   9.10        2010  
8            22  2017   114343   9.05        2010  
9            13  2012   137790   9.05        2010  
10            1  2016   621825   9.03        2010  
...         ...   ...      ...    ...         ...  
2991         26  2005    57573   7.17        2000  
2992          1  1982     5258   7.17        1980  
2993          1  2017     1649   7.17        2010  
2994         35  1983     2996   7.17        1980  
2995         24  2014   911220   7.17        2010  
2996         48  1991     1746   7.17        1990  
2997          1  2013    15645   7.17        2010  
2998          1  2009     5668   7.17        2000  
2999         49  1980     3648   7.17        1980  
3000         12  2016     9884   7.17        2010  

[2930 rows x 8 columns]

Sedaj se lotimo prvega vprašanja: Ali je zlata doba animejev že mimo, ali se šele začenja?

In [49]:
%matplotlib inline
import matplotlib
animeji[animeji.ocena >= 8.0].groupby('desetletje').plot()

ImportError: matplotlib is required for plotting.